functionary/ui/views/functions.py

```diff
@@ -0,0 +1,76 @@
+from django.http import HttpResponse, HttpResponseForbidden, HttpResponseRedirect
+from django.shortcuts import render
+from django.views.generic.detail import DetailView
+from django.views.generic.list import ListView
+
+from core.auth import Permission
+from core.models import Environment, Function, Task
+
+from ..forms.forms import FunctionForm
+
+
+class FunctionListView(ListView):
+    model = Function
+
+    def get_queryset(self):
+        """Filters functions not in the environment then sorts based
+        on package name, then function name."""
+
+        env_id = self.request.session["environment_id"]
+        return (
+            super()
+            .get_queryset()
+            .filter(package__environment__id=env_id)
+            .order_by("package__name", "name")
+        )
+
+
+class FunctionDetailView(DetailView):
+    model = Function
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        form: FunctionForm = FunctionForm(self.get_object())
+        context["form"] = form.render("forms/function_detail.html")
+        return context
+
+
+def execute(request) -> HttpResponse:
+    func: Function = None
+    form: FunctionForm = None
+
+    if request.method == "POST":
+        env: Environment = Environment.objects.get(
+            id=request.session.get("environment_id")
+        )
+        permissions = request.user.environment_permissions(env, inherited=True)
+
+        if request.user.is_superuser or Permission.TASK_CREATE.value in permissions:
```

Comment:
I think you can avoid the `environment_permissions` call above and possibly just change this to be:

```python
if request.user.has_perm(Permission.TASK_CREATE, env):
```

Reply:
Long term, we should build helpers similar to those in the API views so that the proper response is handled for you if this check fails.  Just a note for the future.